llm-examples/falcon/falcon-7b-instruct-inference.py

```diff
@@ -0,0 +1,349 @@
+            "Write a response that appropriately completes the request."
+        )
+
+        return f"""{INTRO_BLURB}
+        {INSTRUCTION_KEY}
+        {instruction}
+        {RESPONSE_KEY}
+        """
+
+    def predict(self, context, model_input):
+        """
+        This method generates prediction for the given input.
+        """
+        prompt = model_input["prompt"][0]
+        temperature = model_input.get("temperature", [1.0])[0]
+        max_new_tokens = model_input.get("max_new_tokens", [100])[0]
+
+        # Build the prompt
+        prompt = self._build_prompt(prompt)
+
+        # Encode the input and generate prediction
+        encoded_input = self.tokenizer.encode(prompt, return_tensors='pt').to('cuda')
+        output = self.model.generate(encoded_input, do_sample=True, temperature=temperature, max_new_tokens=max_new_tokens)
+    
+        # Decode the prediction to text
+        generated_text = self.tokenizer.decode(output[0], skip_special_tokens=True)
+
+        # Removing the prompt from the generated text
+        prompt_length = len(self.tokenizer.encode(prompt, return_tensors='pt')[0])
+        generated_response = self.tokenizer.decode(output[0][prompt_length:], skip_special_tokens=True)
+
+        return [generated_response]
+
+# COMMAND ----------
+
+from mlflow.models.signature import ModelSignature
+from mlflow.types import DataType, Schema, ColSpec
+
+import pandas as pd
+
+# Define input and output schema
+input_schema = Schema([
+    ColSpec(DataType.string, "prompt"), 
+    ColSpec(DataType.double, "temperature"), 
+    ColSpec(DataType.long, "max_new_tokens")])
+output_schema = Schema([ColSpec(DataType.string)])
+signature = ModelSignature(inputs=input_schema, outputs=output_schema)
+
+# Define input example
+input_example=pd.DataFrame({
+            "prompt":["what is ML?"], 
+            "temperature": [0.5],
+            "max_new_tokens": [100]})
+
+# Log the model with its details such as artifacts, pip requirements and input example
+with mlflow.start_run() as run:  
+    mlflow.pyfunc.log_model(
+        "model",
+        python_model=Falcon(),
+        artifacts={'repository' : snapshot_location},
+        pip_requirements=["torch", "transformers", "accelerate", "einops","sentencepiece"],
+        input_example=input_example,
+        signature=signature,
+    )
+
+# COMMAND ----------
+
+# MAGIC %md
+# MAGIC ### Register the model
+
+# COMMAND ----------
+
+# Register model in MLflow Model Registry
+result = mlflow.register_model(
+    "runs:/"+run.info.run_id+"/model",
+    name="falcon-7b-instruct",
+    await_registration_for=600,
+)
+
+# COMMAND ----------
+
+# MAGIC %md
+# MAGIC ### Load the model from model registry
+# MAGIC Assume that the below code is run separately or after the memory cache is cleared.
+# MAGIC
+# MAGIC You can clear the GPU memory by "Detach & re-attach" the notebook.
+
+# COMMAND ----------
+
+# MAGIC %pip install -q -U torch==2.0.1
```

Comment:
DO we need this?

Reply:
Yes, it would require clearing the GPU memory first and thus needs to re-install the python packages.

Reply:
We don't need to re-install the package. We can just cleanup the GPU memory. Something like:
```
from pynvml import *

def print_gpu_utilization():
    nvmlInit()
    handle = nvmlDeviceGetHandleByIndex(0)
    info = nvmlDeviceGetMemoryInfo(handle)
    print(f"GPU memory occupied: {info.used//1024**2} MB.")

print_gpu_utilization()

from numba import cuda
device = cuda.get_current_device()
device.reset()
print_gpu_utilization()
```

Reply:
this is interesting, can you add comments about why we need to clear GPU memory here? what would happen if we don't?

Reply:
In the previoud cells, the model is already loaded into GPU. We need to clean up the memory to load the model from MLflow.

Reply:
Thanks, finally learned how to clear GPU memory effectively! Changed to this now.